storage/ndb/rest-server2/server/src/api_key.cpp

```diff
@@ -26,371 +26,497 @@
 #include <iostream>
 #include <memory>
 #include <openssl/evp.h>
-#include <pthread.h>
 #include <thread>
 #include <unordered_map>
 #include <functional>
 #include <algorithm>
-
-std::shared_ptr<APIKeyCache> apiKeyCache;
+#include <NdbThread.h>
+#include <util/require.h>
+#include <EventLogger.hpp>
+
+//#define DEBUG_AUTH 1
+//#define DEBUG_AUTH_THREAD 1
+//#define DEBUG_AUTH_TIME 1
+//#define DEBUG_AUTH_DBS 1
+
+#ifdef DEBUG_AUTH
+#define DEB_AUTH(arglist) do { g_eventLogger->info arglist ; } while (0)
+#else
+#define DEB_AUTH(arglist) do { } while (0)
+#endif
+
+#ifdef DEBUG_AUTH_THREAD
+#define DEB_AUTH_THREAD(arglist) do { g_eventLogger->info arglist ; } while (0)
+#else
+#define DEB_AUTH_THREAD(arglist) do { } while (0)
+#endif
+
+#ifdef DEBUG_AUTH_TIME
+#define DEB_AUTH_TIME(arglist) do { g_eventLogger->info arglist ; } while (0)
+#else
+#define DEB_AUTH_TIME(arglist) do { } while (0)
+#endif
+
+#ifdef DEBUG_AUTH_DBS
+#define DEB_AUTH_DBS(arglist) do { g_eventLogger->info arglist ; } while (0)
+#else
+#define DEB_AUTH_DBS(arglist) do { } while (0)
+#endif
+
+#ifndef DEBUG_AUTH_THREAD
+#define CLEANUP_SLEEP_TIME 10
+#else
+#define CLEANUP_SLEEP_TIME 1000
+#endif
+extern EventLogger *g_eventLogger;
+
+APIKeyCache *apiKeyCache = nullptr;
 
 std::vector<std::string> split(const std::string &, char);
 RS_Status computeHash(const std::string &unhashed, std::string &hashed);
 
+APIKeyCache* start_api_key_cache() {
+  apiKeyCache = new APIKeyCache();
+  require(apiKeyCache != nullptr);
+  DEB_AUTH(("API Key Cache started: %p", apiKeyCache));
+  return apiKeyCache;
+}
+
+void stop_api_key_cache() {
+  DEB_AUTH(("API Key Cache stopped: %p", apiKeyCache));
+  if (apiKeyCache != nullptr) {
+    delete apiKeyCache;
+    apiKeyCache = nullptr;
+  }
+}
+
+void APIKeyCache::cleanup() {
+  /* Start by waking all threads */
+  DEB_AUTH(("Cleanup started"));
+  NdbMutex_Lock(m_sleepLock);
+  for (int i = 0; i < NUM_API_KEY_CACHES; i++)
+    NdbMutex_Lock(m_rwLock[i]);
+  m_evicted = true;
+  NdbCondition_Broadcast(m_sleepCond);
+  NdbMutex_Unlock(m_sleepLock);
+  for (int i = 0; i < NUM_API_KEY_CACHES; i++)
+    NdbMutex_Unlock(m_rwLock[i]);
+
+  /* Wait for all objects to complete cleanup */
+  for (int i = 0; i < NUM_API_KEY_CACHES; i++) {
+    NdbMutex_Lock(m_rwLock[i]);
+    while (m_key_cache[i].size() > 0) {
+#ifdef DEBUG_AUTH
+      Uint32 key_cache_size = (Uint32)m_key_cache[i].size();
+#endif
+      NdbMutex_Unlock(m_rwLock[i]);
+      DEB_AUTH(("m_key_cache[%d].size() = %u", i, key_cache_size));
```

Comment:
why not 
```
#ifdef DEBUG_AUTH
      Uint32 key_cache_size = (Uint32)m_key_cache[i].size();
      DEB_AUTH(("m_key_cache[%d].size() = %u", i, key_cache_size));
#endif
      NdbMutex_Unlock(m_rwLock[i]);

```

Reply:
Not sure, but pretty sure it will be part of install tarball, above is to avoid holding the mutex while printing, but obviously not important when running in debug, so can change if need be